### ISSUE TITLE
Support for Measured boot with TPM

### ISSUE THREAD

**Issue by TobiPeterG** [nix-community/lanzaboote#348]
To improve the experience for users of encrypted root partitions, measured boot should be supported. This allows the TPM to unlock the root drive, when the kernel, initrd, cmdine and firmware are in a predicted state. The actual state would have to be measured at boot time. systemd-stub supports this, but I'm not sure about Lanzaboote

There are already tools to predict the PCR values, e.g. pcr-oracle, though it's not available for nixos yet
This also supports multiple entries and a secure system boot doesn't depend on secure boot anymore but the actual state of the boot files.

In a perfect world, an encrypted system would use the predictions to auto unlock on boot in case the files haven't been tempered with.

**Comment by petm5** [comment]
I was thinking the current Rust-based installation script could be extended to also generate systemd-pcrlock measurements for the current UKI and systemd-boot binary that launches it. That way `systemd-pcrlock make-policy` could run afterward and predict the PCR 4 value for a TPM policy.

**Comment by nikstur** [comment]
> I was thinking the current Rust-based installation script could be extended to also generate systemd-pcrlock measurements for the current UKI and systemd-boot binary that launches it. That way `systemd-pcrlock make-policy` could run afterward and predict the PCR 4 value for a TPM policy.

Currently working on this

**Comment by fakuivan** [comment]
> I was thinking the current Rust-based installation script could be extended to also generate systemd-pcrlock measurements for the current UKI and systemd-boot binary that launches it. That way `systemd-pcrlock make-policy` could run afterward and predict the PCR 4 value for a TPM policy.

As a side note, `systemd-pcrlock` is not available for TPM2 devices with revision under 1.38 . In my case this includes an Intel 7th gen PTT based fTPM. `systemd-measure sign` works fine with systemd-stub PKIs though, even for signed policies.

**Comment by petm5** [comment]
@fakuivan It looks like the main purpose of `systemd-measure sign` is to restrict TPM-bound secrets to a specific subset of signed boot components, and it doesn't seem to provide much benefit apart from enterprise use cases where you have one Secure Boot keypair and need to break it into subkeys. PCR 11 relies on the bootloader to be "honest" and thus still requires Secure Boot and PCR 7 to be effective. Otherwise the root of trust is broken and a modified boot stub could tamper with the kernel during the time between measuring its signature into the TPM and actually excuting it.

`systemd-pcrlock` can verify the entire boot chain independently from Secure Boot, which is a major advantage for dual-booted systems. Windows updates sometimes change the DBX contents and can unexpectedly lock out disk encryption keys bound to PCR 7.

**Comment by fakuivan** [comment]
@petm5 `systemd-cryptenroll` can bind specific PCRs statically while others using signed policies to unlock LUKS partitions. This is sort of a workaround for "old" hardware (low TPM revisions incopatible with pcrlock) where lower PCRs are practically static due to the lack of firmware updates (except for secure boot db updates).

From `systemd-cryptenroll`s man page:

> Note the difference between --tpm2-pcrs= and --tpm2-public-key-pcrs=: the former binds decryption to the current, specific PCR values; the latter binds decryption to any set of PCR values for which a signature by the specified public key can be provided. The latter is hence more useful in scenarios where software updates shall be possible without losing access to all previously encrypted LUKS2 volumes. Like with --tpm2-pcrs=, names defined in the table above can also be used to specify the registers, for instance --tpm2-public-key-pcrs=boot-loader-code+system-identity.

`systemd-measure sign` is used every time you upgrade the kernel or components commonly bundled with UKIs, while the TPM stays as-is after configured, unless the lower PCR values change.

**Comment by petm5** [comment]
@fakuivan

> `systemd-measure sign` is used every time you upgrade the kernel or components commonly bundled with UKIs, while the TPM stays as-is after configured, unless the lower PCR values change.

This project already signs `systemd-boot` and all UKIs with the user's secure boot keypair. Binding to the value of PCR 7 is enough to pin the enrolled keys and make sure the bootloader, kernel, initrd, etc are signed and can be trusted to auto-unlock LUKS volumes.

If I'm understanding the docs correctly:

- `systemd-measure` calculates the value of PCR 11, and can optionally put it in a signed policy section through `systemd-measure sign`. It does not calculate any other PCR values.
- `systemd-boot` loads a UKI and runs the embedded boot stub.
- The boot stub is responsible for measuring components like the kernel, initrd, etc into PCR 11.
- PCR 11 is designed to complement existing verification. For PCR 11 to be trustworthy, the aforementioned UKI must be verified by a previous stage (Secure Boot and/or `systemd-pcrlock` with a PCR 4 policy).

I can't really see how this is an improvement over plain secure boot, except for cases where you trust the main secure boot signing keys but want to limit access to certain kernel versions.

**Comment by fakuivan** [comment]
@petm5 You're right, I missed the part where you mentioned `systemd-pcrlock` to be used specifically for PCR4 prediction on policy generation. Sorry for the confusion.

**Comment by nikstur** [comment]
Closed by #564
